urlbd77b745af_KokaKiwi_MCLauncher_tgz-pJ8-net_contrapunctus_lzma_BenchmarkJ8/src/SevenZip/Compression/LZMA/Decoder.java

```diff
@@ -0,0 +1,408 @@
+    
+    BitTreeDecoder                          m_PosAlignDecoder     = new BitTreeDecoder(
+                                                                          Base.kNumAlignBits);
+    
+    LenDecoder                              m_LenDecoder          = new LenDecoder();
+    LenDecoder                              m_RepLenDecoder       = new LenDecoder();
+    
+    LiteralDecoder                          m_LiteralDecoder      = new LiteralDecoder();
+    
+    int                                     m_DictionarySize      = -1;
+    int                                     m_DictionarySizeCheck = -1;
+    
+    int                                     m_PosStateMask;
+    
+    public Decoder()
+    {
+        for (int i = 0; i < Base.kNumLenToPosStates; i++)
+        {
+            m_PosSlotDecoder[i] = new BitTreeDecoder(Base.kNumPosSlotBits);
+        }
+    }
+    
+    boolean SetDictionarySize(int dictionarySize)
+    {
+        if (dictionarySize < 0)
+        {
+            return false;
+        }
+        if (m_DictionarySize != dictionarySize)
+        {
+            m_DictionarySize = dictionarySize;
+            m_DictionarySizeCheck = Math.max(m_DictionarySize, 1);
+            m_OutWindow.Create(Math.max(m_DictionarySizeCheck, 1 << 12));
+        }
+        return true;
+    }
+    
+    boolean SetLcLpPb(int lc, int lp, int pb)
+    {
+        if (lc > Base.kNumLitContextBitsMax || lp > 4
+                || pb > Base.kNumPosStatesBitsMax)
+        {
+            return false;
+        }
+        m_LiteralDecoder.Create(lp, lc);
+        final int numPosStates = 1 << pb;
+        m_LenDecoder.Create(numPosStates);
+        m_RepLenDecoder.Create(numPosStates);
+        m_PosStateMask = numPosStates - 1;
+        return true;
+    }
+    
+    void Init() throws IOException
+    {
+        m_OutWindow.Init(false);
+        
+        SevenZip.Compression.RangeCoder.Decoder
+                .InitBitModels(m_IsMatchDecoders);
+        SevenZip.Compression.RangeCoder.Decoder
+                .InitBitModels(m_IsRep0LongDecoders);
+        SevenZip.Compression.RangeCoder.Decoder.InitBitModels(m_IsRepDecoders);
+        SevenZip.Compression.RangeCoder.Decoder
+                .InitBitModels(m_IsRepG0Decoders);
+        SevenZip.Compression.RangeCoder.Decoder
+                .InitBitModels(m_IsRepG1Decoders);
+        SevenZip.Compression.RangeCoder.Decoder
+                .InitBitModels(m_IsRepG2Decoders);
+        SevenZip.Compression.RangeCoder.Decoder.InitBitModels(m_PosDecoders);
+        
+        m_LiteralDecoder.Init();
+        int i;
+        for (i = 0; i < Base.kNumLenToPosStates; i++)
+        {
+            m_PosSlotDecoder[i].Init();
+        }
+        m_LenDecoder.Init();
+        m_RepLenDecoder.Init();
+        m_PosAlignDecoder.Init();
+        m_RangeDecoder.Init();
+    }
+    
+    public boolean Code(java.io.InputStream inStream,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 73 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.